### PR TITLE
Make DUAL a reserved keyword and remove synthetic dual handling

### LINKED ISSUE #19988: Schema engine should not have special handling for `dual`
## Overview

The tablet schema engine has multiple places that treat `dual` as a special, virtual table:

- `go/vt/vttablet/tabletserver/schema/engine.go:274` — a virtual `dual` entry is inserted into `se.tables` at load time
- `go/vt/vttablet/tabletserver/schema/engine.go:537` — the reload path seeds `curTables` with `"dual": true` so it never gets dropped
- `go/vt/vttablet/tabletserver/schema/engine.go:920` — refresh skips `dual` explicitly
- `go/vt/vttablet/tabletserver/schema/engine.go:1045` — `TableCount` subtracts the virtual `dual` entry to report the count of "real" tables

This special-casing is incorrect. `` `dual` `` (backtick-escaped) is a valid user table name in MySQL, and `dual` itself is only treated specially by MySQL's parser as an optional placeholder when there is no `FROM` clause — it is not a schema-level concept. The schema engine should not be tracking a synthetic `dual` entry, and it should not be skipping real user tables named `dual` from refresh/drop logic.

The right place to handle "queries against `dual`" is the planner / query-serving layer, not the schema tracker.

## Proposed direction

- Remove the synthetic `dual` insertion in the schema engine's load path.
- Remove the `dual`-specific carve-outs in the reload, refresh, and `TableCount` paths.
- Verify the planner / `vtgate` side already handles `SELECT ... [FROM dual]` without relying on the schema engine reporting `dual` as a known table; add coverage if needed.
- Add a regression test that creates a real user table named `` `dual` `` and confirms it is tracked, refreshed, and dropped like any other table.

## Context

Spotted in https://github.com/vitessio/vitess/pull/19985#discussion_r3161605091 while reviewing changes that added `TableCount`. Out of scope for that PR — tracked separately so it can be picked up after the other `dual`-specific planner fixes land.

### PR DESCRIPTION
**Pull request by arthurschreiber**
## Description

Makes `DUAL` a reserved keyword in the SQL grammar so that virtual dual (`SELECT 1 FROM DUAL`) and a real table named `` `dual` `` produce distinct ASTs. This matches MySQL behavior where unquoted `DUAL` is a virtual pseudo-table while backtick-quoted `` `dual` `` refers to an actual table.

After this change:
- `SELECT 1` and `SELECT 1 FROM DUAL` → `From: nil` in the AST (virtual dual)
- `` SELECT 1 FROM `dual` `` → `From: [AliasedTableExpr{TableName{"dual"}}]` (real table, backtick-quoted on output)

### What changed

**Parser/Lexer:**
- Added `DUAL` to the keyword table as a reserved keyword
- Added `FROM DUAL` production in `from_opt` that sets `From` to nil
- Removed the lexer special-case hack that intercepted "dual" before keyword lookup

**Formatter:**
- `Select.Format` outputs `from dual` when `From` is nil
- `TableName.Format` no longer special-cases `dual` — real tables named `dual` get backtick-quoted since it's now a reserved keyword

**Normalizer:**
- Simplified `isDual` to check for nil/empty `From` instead of name matching

**Planner:**
- `isOnlyDual` simplified to `len(sel.From) == 0`
- Virtual dual routes (`DualRouting`) no longer register in `SemTable` or appear in `TablesUsed`
- Fixed SQL builder to check `IsDual` before `includeTable` lookup (avoids panic on empty table ID)

Several queries that previously errored with `VT13001: should only be used for single tables` now plan successfully, since dual no longer needs a real table entry in the semantic table.

### Schema-engine and downstream cleanup

The parser change exposes a class of long-standing bugs where a backtick-quoted real `dual` user table was silently bypassed by the schema engine, ACL, vstream copy, lock-tables, vtexplain, and the vtgate planner. With `DUAL` now a reserved keyword, the placeholder no longer reaches name-based lookups and those carve-outs are removable:

- **schema engine** no longer inserts a synthetic `dual` entry on `Open`, no longer seeds `curTables` with `dual` on reload, and no longer skips the `GetTableForPos` cache refresh by name. A real `dual` user table flows through created/altered/dropped notifications like any other table.
- **tablet ACL** no longer skips by name match — placeholder `FROM DUAL` queries now have empty `plan.Permissions` so the per-table ACL loop iterates zero times naturally; backtick-quoted user `dual` tables go through ACL like everything else.
- **vstreamer** `matchTable` no longer filters `dual` from copy.
- **tabletmanager** `LOCK TABLES` no longer skips `dual`.
- **vtexplain** no longer accepts unresolved `dual` references silently.
- **vtgate** `VCursorImpl.FindTableOrVindex` no longer synthesizes a `dual` reference table; the unused `getDualTable` helper is removed. CTE recursive references are now modeled with their actual CTE name and `DualRouting` directly, instead of impersonating a fake `dual` reference table. The `Type==Reference && Name=="dual"` carve-outs in `createRoutingForVTable`, the mirror-rule lookup, and `getOperatorFromAliasedTableExpr` are all gone.
- **schemadiff** no longer pre-populates a placeholder `dual` for view analysis or accepts missing `dual` entities silently.
- FROM-less SELECTs (`SELECT 1`, `SELECT NOW()`, `SELECT … FROM DUAL`) no longer report `main.dual` in `TablesUsed` — they touch no real or virtual table.

Adds regression tests for a real `` `dual` `` user table flowing through schema-engine create/refresh/drop, and for ACL applying to backtick-quoted `` `dual` `` user tables.

## Related Issue(s)

Closes #19988.

## Checklist

- [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
- [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
- [x] Tests were added or are not required
- [x] Did the new or modified tests pass consistently locally and on CI?
- [ ] Documentation was added or is not required

## Deployment Notes

`DUAL` is now a reserved keyword. Queries referencing a real table named `dual` will need to backtick-quote it (`` `dual` ``). This is consistent with MySQL behavior.

Queries that joined `dual` with other tables (e.g. `SELECT 42, id FROM dual, user`) were incorrectly allowed before but will now produce a syntax error. This matches MySQL, where `FROM DUAL` cannot be combined with other tables in a join.

A real user table named `` `dual` `` is now tracked, refreshed, dropped, ACL-checked, vstream-copied, and lock-acquired like any other table — previously it was silently bypassed by every layer that name-matched against the synthetic placeholder.

### AI Disclosure

Most of this was written by Claude Code — I provided direction and design decisions.

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/select_cases.json [hunk 1]
```diff
@@ -5539,24 +5479,7 @@
...
-          "Sharded": false
-        },
-        "FieldQuery": "select 1 as x, (select x from dual where 1 != 1) from dual where 1 != 1",
-        "Query": "select 1 as x, (select x from dual) from dual"
-      },
-      "TablesUsed": [
-        "main.dual"
-      ]
-    }
+    "plan": "column 'x' not found"
```
Comment (arthurschreiber) [inline go/vt/vtgate/planbuilder/testdata/select_cases.json:5538]: Self review: This might be an actual regression. Will need to look into where this is coming from.

File: go/vt/vtgate/semantics/analyzer_test.go [hunk 2]
```diff
@@ -1423,8 +1423,8 @@ func TestNextErrors(t *testing.T) {
 		query, expectedError string
 	}{
 		{
-			query:         "select next 2 values from dual",
-			expectedError: "Table information is not provided in vschema for table `dual`",
```
Comment (arthurschreiber) [inline go/vt/vtgate/semantics/analyzer_test.go:1427]: Self review: I think the original query here was invalid, changing it to `dual` is incorrect. Instead, `FROM DUAL` should not be supported on sequence queries.

File: go/vt/vtgate/executor_select_test.go [hunk 3]
```diff
@@ -1312,7 +1312,11 @@ func TestSelectDual(t *testing.T) {
 
 	_, err = executorExec(ctx, executor, session, "select @@aa.bb from TestUnsharded.dual", nil)
 	require.NoError(t, err)
-	utils.MustMatch(t, wantQueries, lookup.Queries)
```
Comment (arthurschreiber) [inline go/vt/vtgate/executor_select_test.go:1327]: Self review: Here as well, the original query was not correct. `TestUnsharded.dual` refers to an "actual" `dual` table existing in the `TestUnsharded` schema.

File: go/vt/vtgate/semantics/analyzer_test.go (hunk 2 above)
Comment (arthurschreiber) [reply go/vt/vtgate/semantics/analyzer_test.go:1427]: I added a test that covers that `FROM DUAL` is not syntactically valid for sequence queries.

File: go/vt/vtgate/planbuilder/testdata/select_cases.json (hunk 1 above)
Comment (arthurschreiber) [reply go/vt/vtgate/planbuilder/testdata/select_cases.json:5538]: This was fixed in 2e728da14d — the virtual dual subqueries now correctly resolve outer SELECT aliases, so the plan is back to normal (just without `main.dual` in `TablesUsed`).

File: go/vt/vtgate/executor_select_test.go (hunk 3 above)
Comment (arthurschreiber) [reply go/vt/vtgate/executor_select_test.go:1327]: Removed the `TestUnsharded.dual` test case in 6e01cebc42 — it was incorrectly assuming virtual dual could be keyspace-qualified.

**Review by nickvanw — APPROVED** [review]
Looks good to me.

I reviewed the current head closely and ran focused parser, semantics, vtgate, and planbuilder suites,

I did not find a blocker.

One non-blocking testing gap: the old qualified virtual-dual case was correctly removed, but we no longer have a direct positive case proving that a real table named ``dual`` still stays distinct from virtual `DUAL` in planning/execution. A quoted case like ``FROM `dual``` or a qualified quoted form would be good follow-up coverage to keep that split pinned down.

**Review by timvaillancourt — APPROVED** [review]
Makes sense to me 👍

File: go/vt/sqlparser/sql.y [hunk 4]
```diff
@@ -5591,9 +5591,12 @@ col_alias:
 
 from_opt:
   %prec EMPTY_FROM_CLAUSE {
-    $$ = TableExprs{&AliasedTableExpr{Expr:TableName{Name: NewIdentifierCS("dual")}}}
+    $$ = nil
```
Comment (systay) [inline go/vt/sqlparser/sql.y:5592]: ❤️ I've wanted this for far too many years 🥹

File: go/vt/vtgate/planbuilder/testdata/select_cases.json [hunk 5]
```diff
@@ -2844,25 +2832,7 @@
...
-        },
-        "FieldQuery": "select 42, id from dual, `user` where 1 != 1",
-        "Query": "select 42, id from dual, `user`"
-      },
-      "TablesUsed": [
-        "main.dual",
-        "user.user"
-      ]
-    },
+    "plan": "syntax error at position 25",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/select_cases.json:2823]: hey, just flagging that this is a user-visible breaking change — `SELECT 42, id FROM dual, user` used to produce a valid scatter plan and now it's a syntax error. This matches MySQL behavior (which also rejects `FROM DUAL, t`), so it's the right call, but the deployment notes only mention backtick-quoting for real tables named `dual`. Might be worth adding a note that cross-joining with unquoted `DUAL` is no longer valid syntax, since someone out there is probably doing it.

File: go/vt/vtgate/semantics/binder.go [hunk 6]
```diff
@@ -460,9 +460,31 @@ func (b *binder) resolveColumnInScope(current *scope, expr *sqlparser.ColName, a
 		// if we have a failure from uncertain, we matched the column to multiple non-authoritative tables
 		return nil, NotSingleRouteErr{Inner: newAmbiguousColumnError(expr)}
 	}
+	// For scopes with no tables (virtual dual), check if the column matches
+	// a SELECT alias. This allows subqueries like (SELECT x) to reference
+	// aliases from an outer SELECT without a FROM clause.
+	if deps.empty() && len(current.tables) == 0 && expr.Qualifier.IsEmpty() {
```
Comment (systay) [inline go/vt/vtgate/semantics/binder.go]: this is a neat solution for the outer-alias resolution problem, but the trigger condition `len(current.tables) == 0` is broader than "is virtual dual" — it would fire for any scope that somehow ends up with no tables. Right now virtual dual is the only way that happens, but it might be worth either (a) adding a short comment explaining that invariant, or (b) tightening the check (e.g. checking that `current.stmt` is a `*Select` with empty `From`). The `*Select` cast is already there, but the `From` length isn't checked — so if a future change introduces a no-table scope for a different reason, this would silently kick in.

File: go/vt/vttablet/tabletserver/planbuilder/testdata/lock_cases.txt [hunk 7]
```diff
@@ -2,83 +2,47 @@
 "select is_free_lock('foo') from dual"
 {
   "PlanID": "Select",
```
Comment (systay) [inline go/vt/vttablet/tabletserver/planbuilder/testdata/lock_cases.txt:4]: nitpick: plans for `SELECT ... FROM dual` no longer include table-level `Permissions` for the dual table. This is correct — virtual dual shouldn't need permissions — but it's a behavioral change visible to anything inspecting plan permissions at the tablet level. Just noting it here in case there's downstream code that relies on seeing a `dual` permission entry (I don't think there is, but worth a thought).

File: go/vt/sqlparser/ast_format.go [hunk 8]
```diff
@@ -56,12 +56,17 @@ func (node *Select) Format(buf *TrackedBuffer) {
 		buf.literal(SQLCalcFoundRowsStr)
 	}
 
-	buf.astPrintf(node, "%v from ", node.SelectExprs)
+	buf.astPrintf(node, "%v", node.SelectExprs)
```
Comment (systay) [inline go/vt/sqlparser/ast_format.go:60]: nitpick: the string `" from dual"` is lowercase here, relying on `WriteLiteral` to uppercase in canonical mode. This works correctly (verified in tests), but most other places in the formatter write SQL keywords in their canonical case directly. Minor style inconsistency — not blocking, just something that might confuse someone reading the formatter code later.

**Review by systay — COMMENTED** [review]
deep review of #19579 — 4 comments

Nice cleanup overall! The nil-From representation for virtual dual is much cleaner than the string-matching approach. The Append literal fix is a great catch too. A few things caught my eye below.

**Review by systay — APPROVED** [review]
looks good to me — clean simplification of the dual handling. the nil-From representation is a solid improvement over the string-matching approach. just a couple of things to be aware of (see comments), nothing blocking.

File: go/vt/sqlparser/sql.y [hunk 9]
```diff
@@ -5591,9 +5591,12 @@ col_alias:
...
+  }
+| FROM DUAL {
+    $$ = nil
   }
-  | from_clause
+| from_clause
   {
     $$ = $1
   }
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply go/vt/sqlparser/sql.y:5600]: I updated the PR description to be explicit about this.

File: go/vt/vtgate/semantics/binder.go (hunk 6 above)
Comment (arthurschreiber) [reply go/vt/vtgate/semantics/binder.go]: Done!

File: go/vt/vtgate/planbuilder/testdata/select_cases.json (hunk 5 above)
Comment (arthurschreiber) [reply go/vt/vtgate/planbuilder/testdata/select_cases.json:2823]: I updated the PR description to be explicit about this.

File: go/vt/sqlparser/ast_format.go (hunk 8 above)
Comment (arthurschreiber) [reply go/vt/sqlparser/ast_format.go:60]: I'm not sure about this. All the other codepaths in here use `buf.literal` to write strings/keywords, and I updated this code to also use that (instead of `WriteLiteral`).

I couldn't really find the other places that write SQL keywords in uppercase. 🤔

**Comment by arthurschreiber** [comment]
👋 @systay @mattlord @nickvanw @timvaillancourt 

You all reviewed this PR before, but I unfortunately had to make a bunch of changes due to some other places that still relied on the `dual` hacks we had before.

Most importantly, this now also fixes a few places in vttablet that wouldn't have worked right with just the parser changes:
* Tablets no longer store fake `dual` entries in their schema engine.
* Tablets no longer do ACL checks for `dual` tables on "virtual" dual queries (but they do the check if a "real" table named `dual` is queried).

On top of that, I fixed some more cases in the planner that were based on hacky `dual` workarounds (like recursive CTEs).

Can you all take another look at this? 🙇‍♂️

File: go/vt/vtexplain/vtexplain_vttablet.go [hunk 10]
```diff
@@ -643,7 +643,7 @@ func (t *explainTablet) handleSelect(query string) (*sqltypes.Result, error) {
 
 		tableName := sqlparser.String(sqlparser.GetTableName(table.Expr))
 		columns, exists := t.vte.getGlobalTabletEnv().tableColumns[tableName]
-		if !exists && tableName != "" && tableName != "dual" {
+		if !exists && tableName != "" {
 			return nil, fmt.Errorf("unable to resolve table name %s", tableName)
 		}
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply go/vt/vtexplain/vtexplain_vttablet.go:648]: Fixed and covered with a test in https://github.com/vitessio/vitess/pull/19579/commits/7117c7768627a5f6ed6bca7b40b9cc8f12278fad

**Review by mattlord — APPROVED** [review]
I would not call this a blocker due to the practicality of it... but if we look at:
- go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:538-541
- go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go:401-404

After this PR, SELECT ... FROM DUAL and plain no-FROM selects parse with From:nil, but both filter analyzers only reject len(sel.From) > 1 and then index sel.From[0]. A user-provided VStream/VReplication filter like select 1, select * from dual, or a stale unquoted filter for a real `dual` table can now panic during plan construction instead of returning the existing unsupported/invalid filter error. Please require len(sel.From) == 1 before indexing, and add regression tests for no-FROM / FROM DUAL filters in both paths.

This is most definitely an edge case, but probably one worth cleaning up. What do you think? You wouldn't *really* do these in a workflow but probably worth safeguarding against since we're here (we can just return an error if there is no FROM in this context).

Thanks, @arthurschreiber ! ❤️

**Comment by arthurschreiber** [comment]
> This is most definitely an edge case, but probably one worth cleaning up. What do you think?

This was a very good catch! I pushed some more changes to handle this edge case correctly.
